### PR TITLE
Provide a notification when a user disconnects an integration

### PR DESCRIPTION
**Pull request by base10**
* In place for Greenhouse, iCIMS, Jobvite and NetSuite
  * NetSuite doesn't have its disconnect functionality wired in yet,
    so the functionality there is not wrapped with a test
  * I'm not overly worried about that fact since the same functionality
    is well covered with the disconnect feature specs for the other
    three current integrations
* Subsequent integrations will need to be added to INTEGRATION_MAP
  in ConnectionsController
* For: https://trello.com/c/fZ9WQynV/

### CODE REVIEW

File: app/controllers/connections_controller.rb [hunk 1]
```diff
@@ -1,4 +1,11 @@
 class ConnectionsController < ApplicationController
+  INTEGRATION_MAP = {
```
Comment (jferris) [inline app/controllers/connections_controller.rb]: Instead of using a hash for this, what do you think about using an I18n lookup based on the key in the mailer view?

Comment (base10) [inline app/controllers/connections_controller.rb]: Sure, I can do that.

Comment (base10) [inline app/controllers/connections_controller.rb]: Addressed in 9b626cf. NB: I opted to reuse existing translation keys. Let me know if you think we need something purpose specific, even if it's duplicative.

File: app/models/jobvite/client.rb [hunk 2]
```diff
@@ -30,31 +37,50 @@ def self.all(connection)
...
+      attr_reader :connection, :start, :user
 
       def json_response
-        @json_response ||= JSON.parse(RestClient.get(url))
+        @json_response ||= JSON.parse(
+          RestClient.get(url)
+        )
+      end
+
+      def process_errors
```
Comment (base10) [inline app/models/jobvite/client.rb]: Addressed with 46628da

File: app/models/jobvite/client.rb [hunk 3]
```diff
@@ -1,27 +1,34 @@
...
 
-    def initialize(connection)
+    def initialize(connection:, user:)
       @connection = connection
+      @user = user
     end
 
     def recent_hires
-      CandidatePage.all(connection).flat_map { |page| page.candidates }
+      CandidatePage.all(
```
Comment (base10) [inline app/models/jobvite/client.rb]: Addressed with 46628da

File: app/models/jobvite/client.rb (hunk 2 above)
Comment (base10) [inline app/models/jobvite/client.rb]: Addressed with 46628da

File: app/mailers/connection_mailer.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+        "connection_mailer.authentication_notification.subject",
+        integration: @integration
+      )
+    )
+  end
+
+  private
+
+  def map_connection_type_to_integration(connection_type)
+    t("#{connection_type}_connections.edit.title")
```
Comment (jferris) [inline app/mailers/connection_mailer.rb]: It may be better to make a new key instead of using this key, or to rename it to something more generic. I could see somebody changing `edit.title` to something like "Edit NetSuite Settings" without realizing it will also change the text in emails.

File: app/models/jobvite/bulk_import.rb [hunk 5]
```diff
@@ -22,7 +22,7 @@ def new_importer(user)
       Importer.new(
         user,
         connection: user.jobvite_connection,
-        client: Client.new(user.jobvite_connection),
+        client: Client.new(connection: user.jobvite_connection, user: user),
```
Comment (jferris) [inline app/models/jobvite/bulk_import.rb]: Was the Jobvite client also modified? I'm having a hard time figuring out where these parameters would go: https://github.com/namely/connect/blob/nlw-notify-of-lost-sync-credentials/app/models/jobvite/client.rb

File: app/models/net_suite/client.rb [hunk 6]
```diff
@@ -2,14 +2,21 @@ module NetSuite
   class Client
     BASE_URL = "https://api.cloud-elements.com/elements/api-v2"
 
-    def self.from_env
+    def self.from_env(user)
       new(
+        user: user,
```
Comment (jferris) [inline app/models/net_suite/client.rb]: This is only used by `authorize`. Thoughts on passing the user there instead of the initialize?

File: app/models/jobvite/bulk_import.rb (hunk 5 above)
Comment (base10) [inline app/models/jobvite/bulk_import.rb]: Actually, I pulled that back out, so I need to fix this up.

File: spec/mailers/connection_mailer_spec.rb [hunk 7]
```diff
@@ -0,0 +1,46 @@
+require "rails_helper"
+
+describe ConnectionMailer do
+  it "provides integration name in the subject" do
+    connection_type = "icims"
```
Comment (jferris) [inline spec/mailers/connection_mailer_spec.rb]: The set up for these examples looks like it's the same. Thoughts on merging these all into one `it` block?

File: spec/models/icims/authorized_request_spec.rb [hunk 8]
```diff
@@ -106,6 +130,21 @@ def connection
     @connection ||= build(:icims_connection, :connected)
   end
 
+  def user
+    @user ||= @connection.user
```
Comment (jferris) [inline spec/models/icims/authorized_request_spec.rb]: This looks like a reimplementation of `let`. I'd recommend either:

* Using `let` if we planning on going in that direction
* Restructuring this spec file to avoid instance state

File: spec/models/jobvite/client_spec.rb [hunk 9]
```diff
@@ -119,15 +127,58 @@
...
     end
+
+    context "on authentication failure" do
+      it "returns invalid authentication message" do
+        stub_request(:get, "https://api.jobvite.com/api/v2/candidate").
+          with(
+            query: hash_including("api" => "MY_API_KEY", "sc" => "MY_SECRET")
+          ).
+          to_return(
+            body: '{
```
Comment (jferris) [inline spec/models/jobvite/client_spec.rb]: What do you think about using a heredoc here, doc? (har har har)

``` ruby
to_return(
  body: <<-JSON
{...}
  JSON
)
```

File: spec/services/greenhouse/candidates_importer_spec.rb [hunk 10]
```diff
@@ -45,6 +46,22 @@
             candidates_importer.import
           }.to raise_error(Greenhouse::CandidatesImporter::Unauthorized)
         end
+
+        it "sends an invalid authentication message" do
+          allow_any_instance_of(Greenhouse::ValidRequesterPolicy).to(
```
Comment (jferris) [inline spec/services/greenhouse/candidates_importer_spec.rb]: It took me a few scans to figure out this invocation. I'm not sure if it's clearer or if I'm just used to it, but I think it's the parenthesis after `to` that's throwing me off. What do you think of line-splitting after the dot so you don't need parens?

``` ruby
allow_any_instance_of(Greenhouse::ValidRequesterPolicy).
  to receive(:valid?).and_return(false)
```

**Comment by jferris** [comment]
Some high level ideas/questions:

* What do you think about putting a method on `User` to deliver notifications? There's some duplicated boilerplate around setting up the mailer that might be nice to delegate. Not meaningful duplication, but it would still be tidier.
* What happens to these exceptions after the email is raised? Will the user end up seeing a 500 page? How do they fix their credentials once they get messed up?

File: app/models/net_suite/client.rb (hunk 6 above)
Comment (base10) [inline app/models/net_suite/client.rb]: Tried this out, but doesn't seem like a good fit, since ultimately I'm expecting to find `user` on that object when `wrap_response` gets called

File: app/models/jobvite/bulk_import.rb (hunk 5 above)
Comment (base10) [inline app/models/jobvite/bulk_import.rb]: Fixed in e56df47

File: spec/mailers/connection_mailer_spec.rb (hunk 7 above)
Comment (base10) [inline spec/mailers/connection_mailer_spec.rb]: Think I'd feel better about that since there's no way to get around having to trap the exception.

File: app/models/net_suite/client.rb (hunk 6 above)
Comment (jferris) [inline app/models/net_suite/client.rb]: Ah, I see - the two methods looked like one in the diff view.

File: spec/mailers/connection_mailer_spec.rb (hunk 7 above)
Comment (base10) [inline spec/mailers/connection_mailer_spec.rb]: Fixed with da1e471

File: spec/models/icims/authorized_request_spec.rb (hunk 8 above)
Comment (base10) [inline spec/models/icims/authorized_request_spec.rb]: Fixed in `954b17b`, although, if you look at Line 129, we're doing the same sort of thing with connection, which I'm using here. That method is used in a test helper method. Shall I address that, too?

Comment (jferris) [inline spec/models/icims/authorized_request_spec.rb]: Yeah, I think it would be good to restructure to remove the instance state in general.

File: spec/models/jobvite/client_spec.rb (hunk 9 above)
Comment (base10) [inline spec/models/jobvite/client_spec.rb]: Done in cff37a6

File: spec/services/greenhouse/candidates_importer_spec.rb (hunk 10 above)
Comment (base10) [inline spec/services/greenhouse/candidates_importer_spec.rb]: Fixed with 9fc5db8

File: spec/models/icims/authorized_request_spec.rb (hunk 8 above)
Comment (base10) [inline spec/models/icims/authorized_request_spec.rb]: OK, will do.

Comment (base10) [inline spec/models/icims/authorized_request_spec.rb]: Actually, spending a few minutes with unwinding instance state on this test, I agree it's worthwhile, but this seems like a larger task since there's instance state used in several places. I think refactoring this should be a separate PR.

Opened https://trello.com/c/58rvUSyp/ to address this.

File: app/mailers/connection_mailer.rb (hunk 4 above)
Comment (base10) [inline app/mailers/connection_mailer.rb]: Good call. Fixed up in 2daeafd

**Comment by base10** [comment]
@jferris On your [feedback](https://github.com/namely/connect/pull/123#issuecomment-117273718)

1. Good call, I was trying to think of how to centralize this better. I like this approach and will be implementing it.

2. I expect if the user is importing through the imports controller, they'll get a 500. Thinking it through, though, that's something we should rescue and give them instructions on addressing. We should also address this when the background task runs into bad authentication, but that's possibly out of scope here. Would like your thoughts on that.

The way I'd been thinking about a user addressing the issue by deleting the connection and reestablish it for now. We should quickly iterate on this to provide true edit functionality.

**Comment by jferris** [comment]
@base10 Those `User` changes look good to me.

> I expect if the user is importing through the imports controller, they'll get a 500. Thinking it through, though, that's something we should rescue and give them instructions on addressing. ... The way I'd been thinking about a user addressing the issue by deleting the connection and reestablish it for now.

Yeah, I think that would be good to do as part of this story. Even if we just bring them back to the dashboard with a flash message, it would be useful to tell them "you need to disconnect and reconnect" and not just "I'm sorry, but something went wrong."

> We should also address this when the background task runs into bad authentication, but that's possibly out of scope here.

I think we should probably do that now, as the main reason for wanting emails is to know when an automated/background sync fails. However, I think it's fine to break out a new story for that; I just assume it's next on the priority list.

> We should quickly iterate on this to provide true edit functionality.

Yeah, I think it would be good to throw a story in there for that.

**Comment by base10** [comment]
@jferris Will address controller handling. For the other pieces, I've opened up the following Trello cards:

* Edit connections: https://trello.com/c/xrYEiY3u
* Keep automated/background sync running: https://trello.com/c/I6f1verE

**Comment by base10** [comment]
OK, so that was full of discovery.

Jobvite will now present a flash message and return the user to the dashboard in case of authentication error.

The other three controllers aren't fits for this workflow for the following reasons:

* iCIMS and Greenhouse don't import directly. Instead, they're reached by webhooks, so presenting a flash message serves no benefit and indeed, that workflow isn't reachable
  * In this case, I suspect we need a separate card to present an error status to the calling service
* NetSuite promptly enqueues to DelayedJob, so I suspect any exceptions raised by the NetSuite exporter won't be seen by the controller at all

**Comment by jferris** [comment]
Alright, makes sense. After working through these a little more, it seems like the integrations can be divided into two categories:

* Push sync (on-demand, background)
* Pull sync (subscribe and receive notifications for updates)

For push sync, it can be further divided into import/export, but that shouldn't make much difference.

There may be further refactoring we can do in the controllers and service classes to make it easier to reuse the error handling and authentication management. I'll take a look at that shortly.

**Comment by base10** [comment]
@jferris Any further review comments, then?

**Comment by jferris** [comment]
Looks good!

**Comment by base10** [comment]
\o/ Now, to rebase and merge this sucker.
